lib/Sema/TypeCheckAvailability.cpp

```diff
@@ -3254,13 +3254,22 @@ class ExprAvailabilityWalker : public ASTWalker {
         maybeDiagStorageAccess(S->getDecl().getDecl(), S->getSourceRange(), DC);
       }
     }
-    if (auto *RLE = dyn_cast<RegexLiteralExpr>(E)) {
-      // Regex literals require both the Regex<Output> type to be available, as
-      // well as the initializer that is implicitly called.
-      auto Range = RLE->getSourceRange();
-      diagnoseDeclRefAvailability(Context.getRegexDecl(), Range);
-      diagnoseDeclRefAvailability(RLE->getInitializer(), Range);
+
+    if (auto *LE = dyn_cast<LiteralExpr>(E)) {
+      auto Range = LE->getSourceRange();
+      if (auto *RLE = dyn_cast<RegexLiteralExpr>(LE)) {
+        // Regex literals require both the Regex<Output> type to be available,
+        // as well as the initializer that is implicitly called.
+        diagnoseDeclRefAvailability(Context.getRegexDecl(), Range);
```

Comment:
I wonder if this is something that ought to be applied to the result type of a literal in general. e.g for:

```swift
@available(*, unavailable)
struct S: ExpressibleByIntegerLiteral {
  init(integerLiteral value: Int) {}
}
@available(*, unavailable)
typealias IntegerLiteralType = S

let i = 0
```

We're implicitly forming a call to `S(integerLiteral: 0)`, but aren't erroring that it's unavailable.

Reply:
A long term goal of mine is to run availability checks on the transformed AST so that we can catch these issues in a more generalized way. Not sure if that applies directly here but it seems like it might.

Reply:
Yeah these literal cases are tricky because we don't actually form the call in the AST, we emit them directly in SILGen

Reply:
> I wonder if this is something that ought to be applied to the result type of a literal in general.

I'll create an issue to track that case and look at it separately =]

Reply:
@hamishknight @tshortli Issue created #61890